### ISSUE TITLE
2 issues, with custom login on adminAuth

### ISSUE THREAD

**Issue by cloudhack** [node-red/node-red#1461]
### What are the steps to reproduce?
This is a follow up on post at https://groups.google.com/forum/#!topic/node-red/nAS626mPkxo

I have added a small test project that Illustrates the 2 issues at
https://github.com/cloudhack/node-auth-test

### What happens?
Basicly i found 2 issues when implementing a custom passport Strategy.

Hard coding users isn't really liable, so support for a function that validate users on the fly would be better. I've added a sample on how authenticate() in /red/api/auth/users.js to support this

When a passport Strategy doesnt need to redirect/have a callback, authentication is never handled, and the sign in fails
I've added a sample on how line 187 in /red/api/auth/index.js could mediate that problem

### What do you expect to happen?
supplying a function to Users when adding a strategy to adminAuth should be supported.
First passport.authenticate get called, it should beable to signin the user too

### Please tell us about your environment:

- [x] Node-RED version: 0.17.5
- [ ] node.js version: 8.8.1
- [ ] npm version: 5.4.2
- [ ] Platform/OS: linux/Ubuntu
- [ ] Browser: Chrome

**Comment by knolleary** [comment]
Please can you add some more specific details here, or at least point at the relevant parts of your test project?

**Comment by cloudhack** [comment]
I added comments above 2 constants, that control the flow of the code at 
https://github.com/cloudhack/node-auth-test/blob/master/server.js

If you set useUserArray to true, the code will failed without my suggested modifications to  /red/api/auth/users.js
If you set useGoogleID to false, the code will fail without my suggested modifications to /red/api/auth/index.js

If that isnt enough, then i really need a good hint on what info you need from me. Sorry, first time adding an issue to a major project, not used to the etiquette

**Comment by cloudhack** [comment]
Digging around in the node-red code base i saw an "easy" work around the first issue.
I can override the builtin authenticate() in /red/api/auth/users.js by simply adding my own to adminAuth
( but i still think the node-red code should be updated accordingly too )

```
adminAuth.authenticate = function() {
    var _arguments = arguments;
    return when.promise(function(resolve,reject) {
        var username = _arguments[0];
        if (typeof username !== 'string') {
            username = username.username;
        }
        resolve(users[username]);
    });
}
```

**Comment by cloudhack** [comment]
Found one more error. 
When creating strategy buttons, then URL is not relative.
That gives problems when exposing node red though a proxy like nginx 

Either you need to update /red/api/auth/index.js line 92
"prompts":[{type:"button",label:settings.adminAuth.strategy.label, url:"/auth/strategy"}]
to be 
"prompts":[{type:"button",label:settings.adminAuth.strategy.label, url:"auth/strategy"}]

or, update /public/red/red.js line 425
                        var loginButton = $('<a href="#"></a>',{style: "padding: 10px"}).appendTo(row).click(function() {
                            document.location += field.url;
                        });
to

**Comment by knolleary** [comment]
Hi,

please don't use this issue as a catch-all for anything you find. If you find distinct issues, please raise them separately so they can be addressed individually rather than throw it all in one bucket.

For example, the fact the url isn't relative is a nicely self-contained issue that could have a pull-request submitted to fix independent of the others.

**Comment by knolleary** [comment]
Fixed in 0.18 branch
